bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/metrics/MetricsApiTracerAdapter.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.hbase.wrappers.veneer.metrics;
+
+import com.google.api.gax.tracing.ApiTracer;
+import com.google.cloud.bigtable.metrics.RpcMetrics;
+import com.google.cloud.bigtable.metrics.Timer.Context;
+import com.google.common.base.Stopwatch;
+import org.threeten.bp.Duration;
+
+public class MetricsApiTracerAdapter implements ApiTracer {
+
+  private final RpcMetrics rpcMetrics;
+  private final Context operationTimer;
+  private Context rpcTimer;
+
+  private int attemptCount = 0;
+  private Stopwatch attemptTimer;
+  private long attemptResponseCount = 0;
+
+  public MetricsApiTracerAdapter(RpcMetrics rpcMetrics) {
+    this.rpcMetrics = rpcMetrics;
+    operationTimer = rpcMetrics.timeOperation();
+  }
+
+  @Override
+  public Scope inScope() {
+    return new Scope() {
+      @Override
+      public void close() {}
+    };
+  }
+
+  @Override
+  public void operationSucceeded() {
+    operationTimer.close();
+  }
+
+  @Override
+  public void operationCancelled() {
+    operationTimer.close();
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    operationTimer.close();
+  }
+
+  @Override
+  public void connectionSelected(String id) {}
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    if (attemptCount > 0) {
+      rpcMetrics.markRetry();
+    }
+    rpcTimer = rpcMetrics.timeRpc();
+    attemptCount++;
+    attemptResponseCount = 0;
+  }
+
+  @Override
+  public void attemptSucceeded() {
+    rpcTimer.close();
+  }
+
+  @Override
+  public void attemptCancelled() {
+    rpcTimer.close();
+  }
+
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
+    rpcTimer.close();
+  }
+
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable error) {
+    rpcTimer.close();
+    rpcMetrics.markRetriesExhausted();
+  }
+
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
+    rpcTimer.close();
+    rpcMetrics.markFailure();
```

Comment:
This is a little tricky: if retries are disabled, attemptPermanentFailure will never be called which means we will never mark failures for operations. So markFailure needs to move to operationFailed. However from what I understand markFailure() must be exclusive of markRetriesExhausted(). In other words I think that the current behavior is to count retry exhaustion separately from permanent failures.
So I think you need have attemptPermanentFailure and attemptFailedRetriesExhausted set a volatile flag that operationFailed reads to call the appropriate method